// ng2-materialize/dist/sidenav/sidenav-collapsible/sidenav-collapsible.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var sidenav_collapsible_header_component_1 = require("./sidenav-collapsible-header/sidenav-collapsible-header.component");
var MzSidenavCollapsibleComponent = (function () {
    function MzSidenavCollapsibleComponent(changeDetectorRef, renderer) {
        this.changeDetectorRef = changeDetectorRef;
        this.renderer = renderer;
    }
    MzSidenavCollapsibleComponent.prototype.ngAfterViewInit = function () {
        this.initCollapsible();
    };
    MzSidenavCollapsibleComponent.prototype.initCollapsible = function () {
        var _this = this;
        var options = {
            accordion: false,
            onClose: this.onClose,
            onOpen: this.onOpen,
        };
        // need setTimeout otherwise loading directly on the page cause an error
        setTimeout(function () { return _this.renderer.invokeElementMethod($(_this.collapsible.nativeElement), 'collapsible', [options]); });
        this.changeDetectorRef.detectChanges();
    };
    return MzSidenavCollapsibleComponent;
}());
MzSidenavCollapsibleComponent.decorators = [
    { type: core_1.Component, args: [{
                selector: 'mz-sidenav-collapsible',
                template: "<li><ul #collapsible class=\"collapsible collapsible-accordion\"><li><ng-content select=\"mz-sidenav-collapsible-header\"></ng-content><div class=\"collapsible-body\"><ul><ng-content select=\"mz-sidenav-collapsible-content\"></ng-content></ul></div></li></ul></li>",
                styles: [":host /deep/ .collapsible-header{padding:0 32px}:host /deep/ .collapsible-header i{color:rgba(0,0,0,.54)}:host /deep/ .collapsible-body li a{font-weight:initial}"],
            },] },
];
/** @nocollapse */
MzSidenavCollapsibleComponent.ctorParameters = function () { return [
    { type: core_1.ChangeDetectorRef, },
    { type: core_1.Renderer, },
]; };
MzSidenavCollapsibleComponent.propDecorators = {
    'onClose': [{ type: core_1.Input },],
    'onOpen': [{ type: core_1.Input },],
    'collapsible': [{ type: core_1.ViewChild, args: ['collapsible',] },],
    'header': [{ type: core_1.ContentChild, args: [sidenav_collapsible_header_component_1.MzSidenavCollapsibleHeaderComponent,] },],
};
exports.MzSidenavCollapsibleComponent = MzSidenavCollapsibleComponent;
// Declare the tags to avoid error: '<mz-sidenav-collapsible-x>' is not a known element
// https://github.com/angular/angular/issues/11251
// tslint:disable: directive-selector
var MzSidenavCollapsibleContentDirective = (function () {
    function MzSidenavCollapsibleContentDirective() {
    }
    return MzSidenavCollapsibleContentDirective;
}());
MzSidenavCollapsibleContentDirective.decorators = [
    { type: core_1.Directive, args: [{ selector: 'mz-sidenav-collapsible-content' },] },
];
/** @nocollapse */
MzSidenavCollapsibleContentDirective.ctorParameters = function () { return []; };
exports.MzSidenavCollapsibleContentDirective = MzSidenavCollapsibleContentDirective;
