// ng2-materialize/dist/feature-discovery/feature-discovery.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var MzFeatureDiscoveryComponent = (function () {
    function MzFeatureDiscoveryComponent(elementRef) {
        this.elementRef = elementRef;
        this.targetClass = true;
    }
    MzFeatureDiscoveryComponent.prototype.ngAfterViewInit = function () {
        this.target = $(this.elementRef.nativeElement);
    };
    MzFeatureDiscoveryComponent.prototype.close = function () {
        this.target.tapTarget('close');
    };
    MzFeatureDiscoveryComponent.prototype.open = function () {
        this.target.tapTarget('open');
    };
    return MzFeatureDiscoveryComponent;
}());
MzFeatureDiscoveryComponent.decorators = [
    { type: core_1.Component, args: [{
                selector: 'mz-feature-discovery',
                template: "<div class=\"tap-target-content\"><ng-content></ng-content></div>",
                styles: [":host{display:block}"],
            },] },
];
/** @nocollapse */
MzFeatureDiscoveryComponent.ctorParameters = function () { return [
    { type: core_1.ElementRef, },
]; };
MzFeatureDiscoveryComponent.propDecorators = {
    'targetClass': [{ type: core_1.HostBinding, args: ['class.tap-target',] },],
    'targetId': [{ type: core_1.HostBinding, args: ['attr.data-activates',] }, { type: core_1.Input },],
};
exports.MzFeatureDiscoveryComponent = MzFeatureDiscoveryComponent;
